File: app/src/main/java/ru/ls/donkitchen/app/AppModule.java
package ru.ls.donkitchen.app;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.squareup.otto.Bus;
import com.squareup.otto.ThreadEnforcer;

import org.joda.time.DateTime;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit.RestAdapter;
import retrofit.converter.GsonConverter;
import ru.ls.donkitchen.BuildConfig;
import ru.ls.donkitchen.annotation.IOScheduler;
import ru.ls.donkitchen.annotation.UIScheduler;
import ru.ls.donkitchen.rest.Api;
import ru.ls.donkitchen.rest.converter.DateTimeDeserializer;
import ru.ls.donkitchen.util.AsyncBus;
import rx.Scheduler;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Модуль приложения
 *
 * @author Lord (Kuleshov M.V.)
 * @since 01.04.2015
 */
@Module
public class AppModule {
	protected final DonKitchenApplication application;

	public AppModule(DonKitchenApplication application) {
		this.application = application;

		Timber.plant(new Timber.DebugTree());
	}

	@Provides
	@Singleton
	DonKitchenApplication provideApplication() {
		return application;
	}

	@Provides
	@Singleton
	public Bus provideBus() {
		return new AsyncBus(ThreadEnforcer.ANY);
	}

	@Provides
	@UIScheduler
	@Singleton
	public Scheduler provideUiScheduler() {
		return AndroidSchedulers.mainThread();
	}

	@Provides
	@IOScheduler
	@Singleton
	public Scheduler provideIoScheduler() {
		return Schedulers.io();
	}

	@Provides
	@Singleton
	public Gson providerGson() {
		return new GsonBuilder()
				.registerTypeAdapter(DateTime.class, new DateTimeDeserializer())
				.setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
				.create();
	}

	@Provides
	@Singleton
	public Api provideApi(Gson gson) {
		RestAdapter.Builder b = new RestAdapter.Builder();

		b.setConverter(new GsonConverter(gson));

		if (BuildConfig.DEBUG) {
			b.setLogLevel(RestAdapter.LogLevel.FULL);
		} else {
			b.setLogLevel(RestAdapter.LogLevel.FULL);
		}

		b.setRequestInterceptor(request -> {
//			request.addQueryParam("key", BuildConfig.PLACES_API_KEY);
//			request.addQueryParam("sensor", "false");
		});

		RestAdapter restAdapter = b.setEndpoint(BuildConfig.API_URL).build();
		return restAdapter.create(Api.class);
	}
}

File: app/src/main/kotlin/ru/ls/donkitchen/app/AppModule.kt
package ru.ls.donkitchen.app

import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.j256.ormlite.android.apptools.OpenHelperManager
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import ru.ls.donkitchen.BuildConfig
import ru.ls.donkitchen.activity.base.SchedulersFactory
import ru.ls.donkitchen.activity.base.SchedulersFactoryImpl
import ru.ls.donkitchen.data.rest.Api
import ru.ls.donkitchen.data.storage.ormlite.DatabaseHelper
import ru.ls.donkitchen.rest.converter.DateTimeDeserializer
import ru.terrakok.cicerone.Cicerone
import ru.terrakok.cicerone.NavigatorHolder
import ru.terrakok.cicerone.Router
import timber.log.Timber
import java.util.*
import javax.inject.Singleton


@Module
class AppModule(private val application: DonKitchenApplication) {
    init {
        Timber.plant(Timber.DebugTree())
    }

    @Provides
    @Singleton fun provideApplication(): DonKitchenApplication {
        return application
    }

    @Provides
    @Singleton fun provideApi(gson: Gson): Api {
        val b = Retrofit.Builder()

        val clientBuilder = OkHttpClient.Builder()
        val interceptor = HttpLoggingInterceptor()
        if (BuildConfig.DEBUG) {
            interceptor.level = HttpLoggingInterceptor.Level.BODY
        } else {
            interceptor.level = HttpLoggingInterceptor.Level.BASIC
        }
        clientBuilder.addInterceptor(interceptor)

        b.client(clientBuilder.build())
        b.addConverterFactory(GsonConverterFactory.create(gson))
        b.addCallAdapterFactory(RxJava2CallAdapterFactory.create())

        val retrofit = b.baseUrl(BuildConfig.API_URL).build()

        return retrofit.create(Api::class.java)
    }

    @Provides
    @Singleton fun provideGson(): Gson {
        return GsonBuilder()
                .registerTypeAdapter(Date::class.java, DateTimeDeserializer())
                .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
                .create()
    }

    @Provides @Singleton fun provideCicerone(): Cicerone<Router> {
        return Cicerone.create()
    }

    @Provides @Singleton fun provideNavigatorHolder(cicerone: Cicerone<Router>): NavigatorHolder {
        return cicerone.navigatorHolder
    }

    @Provides @Singleton fun provideNavigatorRouter(cicerone: Cicerone<Router>): Router {
        return cicerone.router
    }

    @Provides @Singleton fun provideSchedulersFactory(): SchedulersFactory {
        return SchedulersFactoryImpl()
    }

    @Provides @Singleton fun provideDatabaseHelper(): DatabaseHelper {
        return OpenHelperManager.getHelper(application, DatabaseHelper::class.java)
    }
}
